tests/bundle.test.js

```diff
@@ -0,0 +1,63 @@
+import test from "ava";
+import tokamak from "../src/tokamak.js";
+import { conclude } from "conclure";
+
+test("basic integration test", async (t) => {
+  let requireGraph = {};
+
+  const fileMap = {
+    "file:///file1.js": 'import test from "ava";\nexport default 42;',
+  };
+
+  const loader = {
+    load: function* load(url) {
+      if (!url.startsWith("file://")) {
+        throw new Error(`Don't know how to load ${url}`);
+      }
+
+      requireGraph[url] = {
+        id: url,
+        code: fileMap[url], // this gets transpiled later
+      };
+
+      return requireGraph[url];
+    },
+    isDirectory(url) {
+      return false;
+    },
+    isFile(url) {
+      return true;
+    },
+  };
+
+  const loadModule = tokamak({
+    loader,
+    logger: (level, ...messages) => {
+      console.log(`[${level.toUpperCase()}]`, ...messages);
+    },
+  });
+
+  await conclude(
```

Comment:
Is the conclude call still necessary here? This feels a little awkward

Reply:
May be you should take a look at [Conclure docs](https://github.com/dmaevsky/conclure). Generators completely replace promises, so there should ideally be no async / await anywhere in your code dealing with tokamak.
There are paths to gradually migrate from `async / await` to `* / yield`

`conclude` function is usually only called once at the top level: it is roughly equivalent to `promise.then`, and is NOT an async function

Reply:
Got it! However, I'm a bit confused about how to make sure the test waits for the top-level conclude to complete. Ava seems to have support for async tests, but that obviously doesn't work here. Right now the test works after removing all the awaits, but I think that's because it's not actually doing anything asynchronous right now.

I had a look at the conclure tests to see if there was something that matched what I was trying to do, but didn't see anything:

https://github.com/dmaevsky/conclure/blob/master/tests/conclude.spec.js

Reply:
There are some there. Like this:
```
test.cb('conclude', t => {
  function* myTest() {
    yield Promise.resolve();    // Just insert a tick 
    return 42;
  }

  conclude(myTest(), (error, result) => {
    t.is(error, null);
    t.is(result, 42);
    t.end();
  });
```

Reply:
`test.cb` is your friend :)

Reply:
https://github.com/dmaevsky/conclure/blob/0756ed2bf1f178c5f2bdad25f6e1d1c60aae7816/tests/conclude.spec.js#L21

Reply:
Unfortunately it looks like the latest release of ava removes support for test.cb 😢 https://github.com/avajs/ava/commit/8571b2a7f7ebfefcdc014e7a43b6c874b765b722

But since we're using an older version here, we can probably get away with using it.

Reply:
Updated.

Reply:
Arrrgh, AVA is really getting on my nerves. It is not the first time they are taking all the wrong turns. I am longing to be shown something better. The last time I surveyed other options, I couldn't find anything that would satisfy all my needs:
I had a bad experience with Jest: it is just too slow for some reason.
And nothing else had all the features I needed:
- support for async tests
- nice formatting out of the box
- snapshot testing

But if you have a better alternative, pls let me know.

Otherwise, what would also work is:
```
import { promisify } from 'util';
import { conclude } form 'conclure';
const run = promisify(conclude);  // that's equivalent to the run function in one of my previous comments

test('bla', async t => {
  await run(loadModule(...));
});
```
That's probably more elegant / readable.
I am just a bit allergic to the `async` keyword :)

Reply:
I don't really have any better suggestions, sorry! I use jest for most of my front end stuff, I don't love it but it works well enough. 

I'd probably vote for just leaving things as-is for now: unless there's a compelling reason to upgrade, using an older version of ava (and taking advantage of the features it has to make these types of tests work well) sounds fine.